src/Jackalope/Version/VersionHandler.php

```diff
@@ -245,5 +247,172 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+                continue;
+            }
+
+            $onParentVersion = $property->getDefinition()->getOnParentVersion();
+
+            if ($onParentVersion == OnParentVersionAction::COPY
+                || $onParentVersion == OnParentVersionAction::VERSION
+                || $onParentVersion == OnParentVersionAction::ABORT
+            ) {
+                $property->remove();
+            }
+
+            // TODO handle other onParentVersion cases
+        }
+
+        // TODO handle identifier collisions
+
+        // TODO handle chained versions on restore
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            // TODO handle different onParentVersion values
+
+            $childNodePath = $node->getPath() . '/' . $frozenChildNode->getName();
+            if ($this->session->nodeExists($childNodePath)) {
+                $this->session->removeItem($childNodePath);
+            }
+
+            $this->restoreFromNode($node, $frozenChildNode);
+
+            // TODO remove any node with the same identifier or child identifiers
+        }
+
+        // handle child nodes present on the node but not the frozen node
+        foreach ($node->getNodes() as $childNode) {
+            /** @var NodeInterface $childNode */
+            if ($frozenNode->hasNode($childNode->getName())) {
+                continue;
+            }
+
+            // TODO That's the correct behavior for the OPVs COPY, VERSION and ABORT, handle others as well
+            $childNode->remove();
+        }
+
+        $node->setProperty('jcr:isCheckedOut', false, PropertyType::BOOLEAN, false);
+
+        $this->session->save();
+    }
+
+    /**
+     * @param NodeInterface $sourceNode
+     * @param NodeInterface $destinationNode
+     */
+    private function copyIntoNode(NodeInterface $sourceNode, NodeInterface $destinationNode)
+    {
+        $copiedNode = $destinationNode->addNode($sourceNode->getName(), $sourceNode->getPrimaryNodeType()->getName());
+
+        foreach ($sourceNode->getProperties() as $property) {
+            if ($property->getName() == 'jcr:primaryType') {
+                continue;
+            }
+
+            /** @var PropertyInterface $property */
+            $copiedNode->setProperty($property->getName(), $property->getValue(), $property->getType(), false);
+        }
+
+        foreach ($sourceNode->getNodes() as $childNode) {
+            $this->copyIntoNode($childNode, $copiedNode);
+        }
+    }
+
+    /**
+     * @param NodeInterface $node
+     * @param NodeInterface $frozenNode
+     */
+    private function restoreFromNode(NodeInterface $node, NodeInterface $frozenNode)
```

Comment:
that should be $parentNode and $frozenChildNode i think, to make it obvious. (if we need this, see https://github.com/jackalope/jackalope/pull/289/files#r35429501)